src/backend/cassandra/impl/ExecutionStrategy.h

```diff
@@ -0,0 +1,464 @@
+                throwErrorIfNeeded(res.error());
+            }
+        }
+    }
+
+    /**
+     * @brief Coroutine-based query execution used for reading data.
+     *
+     * Attempts to execute each statement. On any error the whole vector will be
+     * discarded and exception will be thrown.
+     *
+     * @param token Completion token (yield_context)
+     * @param statements Statements to execute
+     * @throw DatabaseTimeout on db error
+     * @return Vector of results
+     */
+    std::vector<ResultType>
+    readEach(
+        CompletionTokenType token,
+        std::vector<StatementType> const& statements)
+    {
+        auto handler = HandlerType{token};
+        auto result = AsyncResultType{handler};
+
+        std::atomic_bool hadError = false;
+        std::atomic_int numOutstanding = statements.size();
+        numReadRequestsOutstanding_ += statements.size();
+
+        auto futures = std::vector<FutureWithCallbackType>{};
+        futures.reserve(numOutstanding);
+
+        // used as the handler for each async statement individually
+        auto executionHandler =
+            [handler, &hadError, &numOutstanding](auto const& res) mutable {
+                if (not res)
+                    hadError = true;
+
+                // when all async operations complete unblock the result
+                if (--numOutstanding == 0)
+                    boost::asio::post(
+                        boost::asio::get_associated_executor(handler),
+                        [handler]() mutable {
+                            handler(boost::system::error_code{});
+                        });
+            };
+
+        std::transform(
+            std::cbegin(statements),
+            std::cend(statements),
+            std::back_inserter(futures),
+            [this, &executionHandler](auto const& statement) {
+                return handle_.get().asyncExecute(statement, executionHandler);
+            });
+
+        // suspend coroutine until completion handler is called
+        result.get();
+
+        numReadRequestsOutstanding_ -= statements.size();
+
+        if (hadError)
+            throw DatabaseTimeout{};
+
+        std::vector<ResultType> results;
+        results.reserve(futures.size());
+
+        // it's safe to call blocking get on futures here as we already
+        // waited for the coroutine to resume above.
+        std::transform(
+            std::make_move_iterator(std::begin(futures)),
+            std::make_move_iterator(std::end(futures)),
+            std::back_inserter(results),
+            [](auto&& future) {
+                auto entry = future.get();
+                auto&& res = entry.value();
+                return std::move(res);
+            });
+
+        assert(futures.size() == statements.size());
+        assert(results.size() == statements.size());
+        return results;
+    }
+
+private:
+    void
+    incrementOutstandingRequestCount()
```

Comment:
I feel we need to update this name, it is only for writing and the name does not indicate it might get blocked.

Reply:
I'll leave this as is for now - this whole strategy should be reworked later (removing yield completely).. let's leave it for then.